File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/values/PlcTIME_OF_DAY.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.spi.values;

import org.apache.plc4x.java.api.types.PlcValueType;
import org.apache.plc4x.java.spi.codegen.WithOption;
import org.apache.plc4x.java.spi.generation.SerializationException;
import org.apache.plc4x.java.spi.generation.WriteBuffer;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.time.LocalTime;

public class PlcTIME_OF_DAY extends PlcIECValue<LocalTime> {

    public static PlcTIME_OF_DAY of(Object value) {
        if (value instanceof PlcTIME_OF_DAY) {
            return (PlcTIME_OF_DAY) value;
        } else if (value instanceof LocalTime) {
            return new PlcTIME_OF_DAY((LocalTime) value);
        } else if (value instanceof Byte) {
            return new PlcTIME_OF_DAY((Byte) value);
        } else if (value instanceof Short) {
            return new PlcTIME_OF_DAY((Short) value);
        } else if (value instanceof Integer) {
            return new PlcTIME_OF_DAY((Integer) value);
        } else if (value instanceof Long) {
            return new PlcTIME_OF_DAY((Long) value);
        } else if (value instanceof Float) {
            return new PlcTIME_OF_DAY((Float) value);
        } else if (value instanceof Double) {
            return new PlcTIME_OF_DAY((Double) value);
        } else if (value instanceof BigInteger) {
            return new PlcTIME_OF_DAY((BigInteger) value);
        } else if (value instanceof BigDecimal) {
            return new PlcTIME_OF_DAY((BigDecimal) value);
        } else {
            return new PlcTIME_OF_DAY(LocalTime.parse(value.toString()));
        }
    }

    public static PlcTIME_OF_DAY ofMillisecondsSinceMidnight(long millisecondsSinceMidnight) {
        return new PlcTIME_OF_DAY(LocalTime.ofNanoOfDay(millisecondsSinceMidnight * 1000_000));
    }

    public PlcTIME_OF_DAY(Byte secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight);
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(Short secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight);
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(Integer secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight);
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(Long secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight);
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(Float secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight.longValue());
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(Double secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight.longValue());
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(BigInteger secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight.longValue());
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(BigDecimal secondsSinceMidnight) {
        this.value = LocalTime.ofSecondOfDay(secondsSinceMidnight.longValue());
        this.isNullable = false;
    }

    public PlcTIME_OF_DAY(LocalTime value) {
        this.value = value;
        this.isNullable = false;
    }

    @Override
    public PlcValueType getPlcValueType() {
        return PlcValueType.TIME_OF_DAY;
    }

    public long getMillisecondsSinceMidnight() {
        return (value.toNanoOfDay() / 1000_000);
    }

    @Override
    public boolean isLong() {
        return true;
    }

    @Override
    public long getLong() {
        return getMillisecondsSinceMidnight();
    }

    @Override
    public boolean isString() {
        return true;
    }

    @Override
    public String getString() {
        return value.toString();
    }

    @Override
    public boolean isTime() {
        return true;
    }

    @Override
    public LocalTime getTime() {
        return value;
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }

    @Override
    public void serialize(WriteBuffer writeBuffer) throws SerializationException {
        String valueString = value.toString();
        writeBuffer.writeString(getClass().getSimpleName(),
            valueString.getBytes(StandardCharsets.UTF_8).length*8,
            valueString, WithOption.WithEncoding(StandardCharsets.UTF_8.name()));
    }

}

File: plc4go/spi/values/PlcTIME_OF_DAY.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package values

import (
	"context"
	"encoding/binary"
	"fmt"
	"time"

	apiValues "github.com/apache/plc4x/plc4go/pkg/api/values"
	"github.com/apache/plc4x/plc4go/spi/utils"
)

type PlcTIME_OF_DAY struct {
	PlcSimpleValueAdapter
	value time.Time
}

func NewPlcTIME_OF_DAY(value any) PlcTIME_OF_DAY {
	var safeValue time.Time
	switch value.(type) {
	case time.Time:
		castedValue := value.(time.Time)
		safeValue = time.Date(0, 0, 0, castedValue.Hour(), castedValue.Minute(), castedValue.Second(),
			castedValue.Nanosecond(), castedValue.Location())
	case uint32:
		// Interpreted as milliseconds since midnight. The offset carries no timezone,
		// so the components are derived arithmetically instead of by resolving it as
		// an instant: going through time.Unix resolves in the host's local zone and
		// shifts GetHours (and GetMillisecondsSinceMidnight) by the local UTC offset.
		castedValue := value.(uint32)
		seconds := castedValue / 1000
		nanoseconds := (castedValue % 1000) * 1000000
		safeValue = time.Date(0, 0, 0, int(seconds/3600), int(seconds%3600/60),
			int(seconds%60), int(nanoseconds), time.UTC)
	}

	return PlcTIME_OF_DAY{
		value: safeValue,
	}
}

func NewPlcTIME_OF_DAYFromMillisecondsSinceMidnight(millisecondsSinceMidnight uint32) PlcTIME_OF_DAY {
	return NewPlcTIME_OF_DAY(millisecondsSinceMidnight)
}

func (m PlcTIME_OF_DAY) IsRaw() bool {
	return true
}

func (m PlcTIME_OF_DAY) GetRaw() []byte {
	theBytes, _ := m.Serialize()
	return theBytes
}

func (m PlcTIME_OF_DAY) GetMillisecondsSinceMidnight() uint32 {
	midnight := time.Date(0, 0, 0, 0, 0, 0, 0, m.value.Location())
	return uint32(m.value.UnixMilli() - midnight.UnixMilli())
}

// GetHours returns the hour of the day: 0 .. 23.
func (m PlcTIME_OF_DAY) GetHours() uint8 {
	return uint8(m.value.Hour())
}

// GetMinutes returns the minute of the hour: 0 .. 59.
func (m PlcTIME_OF_DAY) GetMinutes() uint8 {
	return uint8(m.value.Minute())
}

// GetSeconds returns the second of the minute: 0 .. 59.
func (m PlcTIME_OF_DAY) GetSeconds() uint8 {
	return uint8(m.value.Second())
}

// GetCentiseconds returns the hundredths of a second: 0 .. 99.
//
// Protocols that transmit a time-of-day as separate BCD fields use this as the
// sub-second component (e.g. Schneider UMAS, where TOD is laid out as
// [centiseconds][seconds][minutes][hours]). It truncates rather than rounds, so
// it stays consistent with GetSeconds for any sub-second remainder.
func (m PlcTIME_OF_DAY) GetCentiseconds() uint8 {
	return uint8(m.value.Nanosecond() / 10_000_000)
}

func (m PlcTIME_OF_DAY) IsTime() bool {
	return true
}

func (m PlcTIME_OF_DAY) GetTime() time.Time {
	return m.value
}

func (m PlcTIME_OF_DAY) GetString() string {
	return m.value.Format("15:04:05.000")
}

func (m PlcTIME_OF_DAY) GetPlcValueType() apiValues.PlcValueType {
	return apiValues.TIME_OF_DAY
}

func (m PlcTIME_OF_DAY) Serialize() ([]byte, error) {
	wb := utils.NewWriteBufferByteBased(utils.WithByteOrderForByteBasedBuffer(binary.BigEndian))
	if err := m.SerializeWithWriteBuffer(context.Background(), wb); err != nil {
		return nil, err
	}
	return wb.GetBytes(), nil
}

func (m PlcTIME_OF_DAY) SerializeWithWriteBuffer(ctx context.Context, writeBuffer utils.WriteBuffer) error {
	return writeBuffer.WriteString("PlcTIME_OF_DAY", uint32(len([]rune(m.GetString()))*8), m.GetString())
}

func (m PlcTIME_OF_DAY) String() string {
	return fmt.Sprintf("%s(%dbit):%v", m.GetPlcValueType(), uint32(len([]rune(m.GetString()))*8), m.value)
}
